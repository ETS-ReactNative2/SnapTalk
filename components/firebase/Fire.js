import FirebaseKeys from './Config';
import firebase from 'firebase';
require('firebase/firestore');

class Fire {

  constructor() {
    firebase.initializeApp(FirebaseKeys);
  }

  addPost = async ({ text, localUri, postKey }) => {
    //Uploading picture to database
    const remoteUri = await this.uploadPhotoAsync(localUri, 'photos/' + this.uid + '/' + Date.now());

    //Getting User from database
    const user = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Pointer to the name field of the user in the database
    const fieldPathName = new firebase.firestore.FieldPath('name');

    //Getting User from database again (because cannot use the same variable)
    const userAgain = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Pointer to the profile picture field of the user in the database
    const fieldPathProfilePicture = new firebase.firestore.FieldPath('profilePicture');

    //Creating a new post in the database
    return new Promise((res, rej) => {
      this.firestore
        .collection('posts')
        .add({
          text,
          uid: this.uid,
          postKey: '',
          timestamp: this.timestamp,
          image: remoteUri,
          username: user.get(fieldPathName),
          avatar: userAgain.get(fieldPathProfilePicture),
          listOfComments: [],
          nbOfComments: 0,
          listOfLikes: [],
        })
        .then(ref => {
          res(ref);
        })
        .catch(error => {
          rej(error);
        });
    });
  };

  updateUserLikedList = async postId => {
    //Getting post from the database
    let db = this.firestore.collection('posts').doc(postId);

    //Getting post from the database
    const post = await firebase
      .firestore()
      .collection('posts')
      .doc(postId)
      .get();

    //Creating pointer to the list Of likes of the post in the database
    const fieldPathListOfLikes = new firebase.firestore.FieldPath('listOfLikes');

    //Getting the list of likes of the post from the database
    const arrayOfLikes = await post.get(fieldPathListOfLikes);

    if (arrayOfLikes.includes(this.uid)) {
      db.update({
        listOfLikes: firebase.firestore.FieldValue.arrayRemove(this.uid),
      });
    }
    else {
      db.update({
        listOfLikes: firebase.firestore.FieldValue.arrayUnion(this.uid),
      });
    }
  };

  addPostKey = async postId => {
    //Getting post from database
    let dbPost = this.firestore.collection('posts').doc(postId);

    if (postId) {
      dbPost.update({
        postKey: postId,
      });
    }
  };

  updatePostList = async postId => {
    //Getting user from the database
    let db = this.firestore.collection('users').doc(this.uid);

    //Getting user from the database
    const user = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Creating pointer to the list of posts field in the database
    const fieldPathListOfPosts = new firebase.firestore.FieldPath('listOfPosts');

    if (postId) {
      db.update({
        listOfPosts: firebase.firestore.FieldValue.arrayUnion(postId),
      });

      const nbOfPosts = (await user.get(fieldPathListOfPosts).length) + 1;

      db.set({ nbOfPosts: nbOfPosts }, { merge: true });
    }
  };

  addComment = async ({ text, postKey }) => {
    //Getting user from database
    const user = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Creating a pointer to name field of the user in the database
    const fieldPathName = new firebase.firestore.FieldPath('name');

    //Getting User from database again (because cannot use the same variable)
    const userAgain = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Creating a pointer to the profile picture field in the database
    const fieldPathProfilePicture = new firebase.firestore.FieldPath('profilePicture');

    return new Promise((res, rej) => {
      this.firestore
        .collection('comments')
        .add({
          comment: text,
          uid: this.uid,
          timestamp: this.timestamp,
          username: user.get(fieldPathName),
          postKey: postKey,
          commentKey: '',
          avatar: userAgain.get(fieldPathProfilePicture),
        })
        .then(ref => {
          res(ref);
        })
        .catch(error => {
          rej(error);
        });
    });
  };

  addCommentKey = async commentId => {
    //Getting comments from database
    let dbComment = this.firestore.collection('comments').doc(commentId);

    if (commentId) {
      dbComment.update({
        commentKey: commentId,
      });
    }
  };

  updateCommentList = async ({ commentId, postId }) => {
    //Getting user from database
    let dbUser = this.firestore.collection('users').doc(this.uid);

    //Getting post from database
    let dbPost = this.firestore.collection('posts').doc(postId);

    //Getting post again from database
    const post = await firebase
    .firestore()
    .collection('posts')
    .doc(postId)
    .get();

    //Getting user again from database
    const user = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Creating pointer to the list of comments field in the database
    const fieldPathListOfComments = new firebase.firestore.FieldPath('listOfComments');

    if (commentId) {
      dbUser.update({
        listOfComments: firebase.firestore.FieldValue.arrayUnion(commentId),
      });

      const nbOfComments = (await user.get(fieldPathListOfComments).length) + 1;

      dbUser.set({ nbOfComments: nbOfComments }, { merge: true });
    }

    if (postId) {
      dbPost.update({
        listOfComments: firebase.firestore.FieldValue.arrayUnion(commentId)
      });

      const nbOfComments = (await post.get(fieldPathListOfComments).length) + 1;

      dbPost.set({ nbOfComments: nbOfComments }, { merge: true });
    }
  };

  modifyComment = async (commentKey, newComment) => {
    //Getting comments from database
    let dbComment = this.firestore.collection('comments').doc(commentKey);

    if (commentKey) {
      dbComment.update({
        comment: newComment,
      });
    }
  };

  deleteComment = async (commentId, postId) => {

    //Getting comment from database
    let dbComments = this.firestore.collection('comments').doc(commentId);

    //Getting user from database
    let dbUser = this.firestore.collection('users').doc(this.uid);

    //Getting post from database
    let dbPost = this.firestore.collection('posts').doc(postId);

    //Getting post again from database
    const post = await firebase
    .firestore()
    .collection('posts')
    .doc(postId)
    .get();

    //Getting user again from database
    const user = await firebase
      .firestore()
      .collection('users')
      .doc(this.uid)
      .get();

    //Creating pointer to the list of comments field in the database
    const fieldPathListOfComments = new firebase.firestore.FieldPath('listOfComments');

    if (commentId) {

      dbComments.delete();

      dbUser.update({
        listOfComments: firebase.firestore.FieldValue.arrayRemove(commentId)
      });

      const nbOfComments = (await user.get(fieldPathListOfComments).length) - 1;

      dbUser.set({ nbOfComments: nbOfComments }, { merge: true });
    }

    if (postId) {
      dbPost.update({
        listOfComments: firebase.firestore.FieldValue.arrayRemove(commentId),
      });

      const nbOfComments = (await post.get(fieldPathListOfComments).length) - 1;

      dbPost.set({ nbOfComments: nbOfComments }, { merge: true });
    }
  };



  createUser = async user => {
    let remoteAvatarUri = null;

    try {

      if (!user.name) {
        throw new Error('Username was not entered.');
      }

      await this.firestore
        .collection('users')
        .get()
        .then(snapshot => {
          snapshot.forEach(doc => {
            if (user.name == doc.data().name) {
              throw new Error('Username already taken.');
            }
          });
        });

      await firebase
        .auth()
        .createUserWithEmailAndPassword(user.email, user.password);

      let db = this.firestore.collection('users').doc(this.uid);

      db.set({
        name: user.name,
        email: user.email,
        listOfFollowers: [],
        listOfFollowing: [],
        listOfPosts: [],
        listOfComments: [],
        nbOfPosts: 0,
        nbOfComments: 0,
        profilePicture: remoteAvatarUri
      });

      if (user.avatar) {
        remoteAvatarUri = await this.uploadPhotoAsync(
          user.avatar,
          'avatar/' + this.uid,
        );
        db.set({ profilePicture: remoteAvatarUri }, { merge: true });
      }

    }
    catch (error) {
      alert('Error: ' + error.message);
    }
  };

  uploadPhotoAsync = async (uri, filename) => {
    //Uploading picture to firebase storage
    return new Promise(async (res, rej) => {
      const response = await fetch(uri);
      const file = await response.blob();

      let upload = firebase
        .storage()
        .ref(filename)
        .put(file);

      upload.on(
        'state_changed',
        snapshot => { },
        err => {
          rej(err);
        },
        async () => {
          const url = await upload.snapshot.ref.getDownloadURL();
          res(url);
        },
      );
    });
  };

  get firestore() {
    return firebase.firestore();
  }

  get uid() {
    return (firebase.auth().currentUser || {}).uid;
  }

  get timestamp() {
    return Date.now();
  }

}

Fire.shared = new Fire();
export default Fire;