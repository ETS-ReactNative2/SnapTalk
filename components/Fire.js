import FirebaseKeys from "./config";
import firebase from 'firebase';
require("firebase/firestore");

class Fire {

    constructor() {
        firebase.initializeApp(FirebaseKeys);
    }


    addPost = async ({ text, localUri, postKey }) => {
        const remoteUri = await this.uploadPhotoAsync(localUri, 'photos/' + this.uid + '/' + Date.now());

        const user = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathName = new firebase.firestore.FieldPath('name');

        const userAgain = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathProfilePicture = new firebase.firestore.FieldPath('profilePicture');

        return new Promise((res, rej) => {
            this.firestore.collection("posts").add({
                text,
                uid: this.uid,
                postKey: '',
                timestamp: this.timestamp,
                image: remoteUri,
                username: user.get(fieldPathName),
                avatar: userAgain.get(fieldPathProfilePicture),
                listOfComments: [],
                nbOfComments: 0,
                listOfLikes: [],
                nbOfLikes: 0,
            })
                .then(ref => {
                    res(ref);
                })
                .catch(error => {
                    rej(error);
                })
        })
    };

    updateUserLikedList = async (postId) => {
        let db = this.firestore.collection("posts").doc(postId);

        const post = await firebase.firestore().collection("posts").doc(postId).get();
        const fieldPathListOfLikes = new firebase.firestore.FieldPath('listOfLikes');
        const arrayOfLikes = await post.get(fieldPathListOfLikes)

        if (arrayOfLikes.includes(this.uid)) {
            db.update({
                listOfLikes: firebase.firestore.FieldValue.arrayRemove(this.uid),
            })
        }
        else {
            db.update({
                listOfLikes: firebase.firestore.FieldValue.arrayUnion(this.uid),
            })
        }
    }



    addPostKey = async (postId) => {

        let dbPost = this.firestore.collection("posts").doc(postId);

        if (postId) {
            dbPost.update({
                postKey: postId
            })
        }
    };

    updatePostList = async (postId) => {
        let db = this.firestore.collection("users").doc(this.uid);

        const user = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathListOfPosts = new firebase.firestore.FieldPath('listOfPosts');

        if (postId) {
            db.update({
                listOfPosts: firebase.firestore.FieldValue.arrayUnion(postId)
            })

            const nbOfPosts = await user.get(fieldPathListOfPosts).length + 1;

            db.set(
                { nbOfPosts: nbOfPosts }, { merge: true }
            )
        }
    }

    addComment = async ({ text, postKey }) => {
        const user = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathName = new firebase.firestore.FieldPath('name');

        const userAgain = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathProfilePicture = new firebase.firestore.FieldPath('profilePicture');

        return new Promise((res, rej) => {
            this.firestore.collection("comments").add({
                comment: text,
                uid: this.uid,
                timestamp: this.timestamp,
                username: user.get(fieldPathName),
                postKey: postKey,
                commentKey: '',
                avatar: userAgain.get(fieldPathProfilePicture)
            })
                .then(ref => {
                    res(ref);
                })
                .catch(error => {
                    rej(error);
                })
        })
    };

    addCommentKey = async (commentId) => {

        let dbComment = this.firestore.collection("comments").doc(commentId);

        if (commentId) {
            dbComment.update({
                commentKey: commentId
            })
        }
    };

    updateCommentList = async ({ commentId, postId }) => {
        let dbUser = this.firestore.collection("users").doc(this.uid);

        let dbPost = this.firestore.collection("posts").doc(postId);

        const user = await firebase.firestore().collection("users").doc(this.uid).get();

        const fieldPathListOfComments = new firebase.firestore.FieldPath('listOfComments');

        if (commentId) {
            dbUser.update({
                listOfComments: firebase.firestore.FieldValue.arrayUnion(commentId)
            })

            const nbOfComments = await user.get(fieldPathListOfComments).length + 1;

            dbUser.set(
                { nbOfComments: nbOfComments }, { merge: true }
            )
        }

        if (postId) {
            dbPost.update({
                listOfComments: firebase.firestore.FieldValue.arrayUnion(commentId)
            })

            const nbOfComments = await user.get(fieldPathListOfComments).length + 1;

            dbPost.set(
                { nbOfComments: nbOfComments }, { merge: true }
            )
        }
    }

    createUser = async user => {
        let remoteAvatarUri = null

        try {
            if (!user.name) {
                throw new Error("Username was not entered.")
            }

            await this.firestore
                .collection("users")
                .get()
                .then(snapshot => {
                    snapshot.forEach(doc => {
                        if (user.name == doc.data().name) {
                            throw new Error("Username already taken.")
                        }
                    })
                })

            await firebase.auth().createUserWithEmailAndPassword(user.email, user.password);

            let db = this.firestore.collection("users").doc(this.uid);

            db.set({
                name: user.name,
                email: user.email,
                listOfFollowers: [],
                listOfFollowing: [],
                listOfPosts: [],
                listOfComments: [],
                nbOfPosts: 0,
                nbOfComments: 0,
                profilePicture: remoteAvatarUri
            });

            if (user.avatar) {
                remoteAvatarUri = await this.uploadPhotoAsync(user.avatar, 'avatar/' + this.uid);
                db.set({ profilePicture: remoteAvatarUri }, { merge: true })
            }

        } catch (error) {
            alert("Error: " + error.message);
        }
    }

    uploadPhotoAsync = async (uri, filename) => {

        return new Promise(async (res, rej) => {
            const response = await fetch(uri);
            const file = await response.blob();

            let upload = firebase
                .storage()
                .ref(filename)
                .put(file);

            upload.on("state_changed", snapshot => { }, err => {
                rej(err);
            },
                async () => {
                    const url = await upload.snapshot.ref.getDownloadURL();
                    res(url);
                }
            );
        });
    };


    get firestore() {
        return firebase.firestore();
    }

    get uid() {
        return (firebase.auth().currentUser || {}).uid;
    }

    get timestamp() {
        return Date.now();
    }
}


Fire.shared = new Fire();
export default Fire;