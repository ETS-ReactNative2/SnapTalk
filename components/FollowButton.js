import React, { Component } from 'react';
import { Alert, Button, TouchableHighlight, Text, StyleSheet} from 'react-native';
import { Avatar } from 'react-native-elements';
import Fire from './Fire';

const firebase = require('firebase');
require("firebase/firestore");

export default class FollowButton extends Component {

    state = {

    followClick: false,
    isFollowing: false,

    user : {},
    targetUser : {},
    toFollowedUserId:{},


    }

    componentDidMount() {


     this.unsubscribe = Fire.shared.firestore
          .collection("users")
          .get()
          .then(snapshot => {
                  snapshot.forEach(doc => {
                       if (doc.id == this.props.loggedUserUID) {
                           this.setState({user: doc.data()});
                        }
                        if(doc.data().name == this.props.userToFollow){
                              this.setState({toFollowedUserId: doc.id});
                              this.setState({targetUser: doc.data()});
                        }
                    })
              })


              this.unsubscribe = Fire.shared.firestore
                        .collection("users")
                        .get()
                        .then(snapshot => {
                                snapshot.forEach(doc => {
                                     if (doc.id == this.props.loggedUserUID) {

                                          doc.data().listOfFollowing.forEach(followerName => {
                                                    console.log(followerName)

                                                    if(this.state.targetUser.name == followerName){
                                                        this.setState({followClick: true});
                                                        this.setState({isFollowing: true});
                                                    }
                                              }
                                          )
                                      }
                                  })
                            })
      };


    followAction = () => {


    let followingRef = Fire.shared.firestore.collection('users').doc(this.props.loggedUserUID);

    let followerRef = Fire.shared.firestore.collection('users').doc(this.state.toFollowedUserId);

    if(this.state.followClick == false || this.state.isFollowing == false){

        this.setState({followClick: !this.state.followClick, isFollowing: !this.state.isFollowing})

                // add a new Following to the "listOfFollwing" array of the current
                let arrUnionFollowing = followingRef.update({
                  listOfFollowing: firebase.firestore.FieldValue.arrayUnion(this.props.userToFollow)
                });

                // add a new Follower to the "listOfFollowing" array of the other user
                let arrUnionFollower = followerRef.update({
                  listOfFollowers: firebase.firestore.FieldValue.arrayUnion(this.state.user.name)
                });
    }

    if(this.state.followClick == true || this.state.isFollowing == true){

            this.setState({followClick: !this.state.followClick, isFollowing: !this.state.isFollowing})

            let arrRemoveFollowing = followingRef.update({
              listOfFollowing: firebase.firestore.FieldValue.arrayRemove(this.props.userToFollow)
            });

            let arrRemoveFollower = followerRef.update({
              listOfFollowers: firebase.firestore.FieldValue.arrayRemove(this.state.user.name)
            });

        }




    }

  render() {

        const buttonColor = this.state.followClick ? '#68a0cf' : 'white'
        const followState = this.state.isFollowing ? 'Unfollow' : 'Follow'

    return (
        <TouchableHighlight
            style={{width: 70, height: 20, backgroundColor:buttonColor, borderRadius:5, borderWidth: 1, borderColor: 'black',  color: "#52575D", fontFamily: "HelveticaNeue", fontSize: 18,}}
            onPress={this.followAction}
        >
        <Text style = {{textAlign: 'center'}}>{followState}</Text>
        </TouchableHighlight>
    );
  }
}
