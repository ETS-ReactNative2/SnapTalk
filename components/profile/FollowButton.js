import React, { Component } from 'react';
import { TouchableHighlight, Text } from 'react-native';
import Fire from '../firebase/Fire';

const firebase = require('firebase');
require('firebase/firestore');

export default class FollowButton extends Component {

  state = {
    followClick: false,
    isFollowing: false,
    user: {},
    targetUser: {},
    toFollowedUserId: {},
  };

  componentDidMount() {
    //Getting user from database
    this.unsubscribe = Fire.shared.firestore
      .collection('users')
      .get()
      .then(snapshot => {
        snapshot.forEach(doc => {
          if (doc.id == this.props.loggedUserUID) {
            this.setState({ user: doc.data() });
          }
          if (doc.data().uid == this.props.userToFollow) {
            this.setState({ toFollowedUserId: doc.id });
            this.setState({ targetUser: doc.data() });
          }
        });
      });

    this.unsubscribe = Fire.shared.firestore
      .collection('users')
      .get()
      .then(snapshot => {
        snapshot.forEach(doc => {
          if (doc.id == this.props.loggedUserUID) {
            doc.data().listOfFollowing.forEach(followerName => {
              console.log(followerName);

              if (this.state.targetUser.uid == followerName) {
                this.setState({ followClick: true });
                this.setState({ isFollowing: true });
              }
            });
          }
        });
      });
  }

  followAction = () => {
    let followingRef = Fire.shared.firestore
      .collection('users')
      .doc(this.props.loggedUserUID);

    let followerRef = Fire.shared.firestore
      .collection('users')
      .doc(this.state.toFollowedUserId);

    if (this.state.followClick == false || this.state.isFollowing == false) {
      this.setState({
        followClick: !this.state.followClick,
        isFollowing: !this.state.isFollowing,
      });

      // add a new Following to the "listOfFollwing" array of the current user
      let arrUnionFollowing = followingRef.update({
        listOfFollowing: firebase.firestore.FieldValue.arrayUnion(
          this.props.userToFollow,
        ),
      });

      // add a new Follower to the "listOfFollowing" array of the other user
      let arrUnionFollower = followerRef.update({
        listOfFollowers: firebase.firestore.FieldValue.arrayUnion(
          this.state.user.uid,
        ),
      });

      let arrUnionFollowerNotif = followerRef.update({
        listOfNotif: firebase.firestore.FieldValue.arrayUnion(
          this.state.user.name + " is following you!",
        ),
      });

    }

    if (this.state.followClick == true || this.state.isFollowing == true) {
      this.setState({
        followClick: !this.state.followClick,
        isFollowing: !this.state.isFollowing,
      });

      let arrRemoveFollowing = followingRef.update({
        listOfFollowing: firebase.firestore.FieldValue.arrayRemove(
          this.props.userToFollow,
        ),
      });

      let arrRemoveFollower = followerRef.update({
        listOfFollowers: firebase.firestore.FieldValue.arrayRemove(
          this.state.user.uid,
        ),
      });
    }
  };

  render() {
    const buttonColor = this.state.followClick ? '#68a0cf' : 'white';
    const followState = this.state.isFollowing ? 'Unfollow' : 'Follow';

    return (
      <TouchableHighlight
        style={{
          width: 70,
          height: 20,
          backgroundColor: buttonColor,
          borderRadius: 5,
          borderWidth: 1,
          borderColor: 'black',
          color: '#52575D',
          fontFamily: 'HelveticaNeue',
          fontSize: 1
        }}
        onPress={this.followAction}>
        <Text style={{ textAlign: 'center' }}>{followState}</Text>
      </TouchableHighlight>
    );
  }

}
