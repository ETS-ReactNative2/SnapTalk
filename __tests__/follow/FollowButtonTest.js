import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import FollowButton from '../../components/profile/FollowButton'
import Fire from '../../components/firebase/Fire'

describe('FollowButton Test', () => {
    let followButtonSnapshot = renderer.create(<FollowButton />).toJSON();

    test('FollowButton Render Test', () => {
        expect(followButtonSnapshot).toMatchSnapshot();
    })
})