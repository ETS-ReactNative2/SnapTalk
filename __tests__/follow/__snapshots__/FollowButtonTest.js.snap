// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowButton Test FollowButton Render Test 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderColor": "black",
      "borderRadius": 5,
      "borderWidth": 1,
      "height": 20,
      "width": 70,
    }
  }
>
  <Text
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Follow
  </Text>
</View>
`;
